src/hotspot/share/logging/logAsyncWriter.hpp

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+#ifndef SHARE_LOG_ASYNC_WTRITER_HPP
+#define SHARE_LOG_ASYNC_WTRITER_HPP
+#include "logging/log.hpp"
+#include "logging/logDecorations.hpp"
+#include "logging/logFileOutput.hpp"
+#include "logging/logMessageBuffer.hpp"
+#include "memory/resourceArea.hpp"
+#include "runtime/nonJavaThread.hpp"
+#include "utilities/hashtable.hpp"
+#include "utilities/linkedlist.hpp"
+
```

Comment:
Could you write a summary of operation comment explaining how async mode works e.g.  details on initialization and teardown, how the thread interacts with writers etc. Thanks.

Reply:
How about this?

`initialize()` is called once when JVM is initialized. It creates and initialize the singleton instance of AsyncLogWriter. Once async logging is established, there's no way to turn it off. 

`instance()` is MT-safe and returns the pointer of the singleton instance if and only if async mode is on and has well initialized.  Clients can use its return value to determine async logging is established or not. 

The basic operation of AsyncLogWriter is `enqueue`. Two overloading versions of it are provided to match `LogOutput::write`. To support async logging,  derived classes of LogOutput can invoke the corresponding AsyncLogWriter::enqueue in write() and return -1.  AsyncLogWriter is responsible of copying both decorations and cstr messages and managing their lifecycle. 

The static member function `flush`is designated to flush out all pending messages when JVM is terminating or aborting. In normal JVM termination, flush() is invoked in `LogConfiguration::finalize()`.  In abortion situation, `flush()` is invoked in `os::shutdown()`.  `flush` is MT-safe and can be invoked arbitrary times.  It is no-op if asynclogging is not established.